proxy/coco/src/peer/subroutines.rs

```diff
@@ -0,0 +1,280 @@
+        }
+
+        loop {
+            match self.inputs.poll_next_unpin(cx) {
+                Poll::Ready(Some(input)) => {
+                    log::debug!("handling subroutine input: {:?}", input);
+
+                    if let Some(event) = Event::maybe_from(&input) {
+                        // Ignore if there are no subscribers.
+                        self.subscriber.send(event).ok();
+                    }
+
+                    for cmd in self.run_state.transition(input) {
+                        let task = match cmd {
+                            Command::Announce => SpawnAbortable::new(announce(
+                                self.state.clone(),
+                                self.store.clone(),
+                                self.input_sender.clone(),
+                            )),
+                            Command::Control(control_command) => match control_command {
+                                ControlCommand::Respond(respond_command) => {
+                                    SpawnAbortable::new(control_respond(respond_command))
+                                },
+                            },
+                            Command::SyncPeer(peer_id) => SpawnAbortable::new(sync(
+                                self.state.clone(),
+                                peer_id.clone(),
+                                self.input_sender.clone(),
+                            )),
+                            Command::StartSyncTimeout(sync_period) => SpawnAbortable::new(
+                                start_sync_timeout(sync_period, self.input_sender.clone()),
+                            ),
+                            Command::Request(RequestCommand::Query(urn)) => SpawnAbortable::new(
+                                query(urn, self.state.clone(), self.input_sender.clone()),
+                            ),
+                            Command::Request(RequestCommand::Clone(url)) => SpawnAbortable::new(
+                                clone(url, self.state.clone(), self.input_sender.clone()),
+                            ),
+                        };
+
+                        self.pending_tasks.push(task);
+                    }
+                },
+                Poll::Ready(None) => return Poll::Ready(Ok(())),
+                Poll::Pending => return Poll::Pending,
+            }
+        }
+    }
+}
+
+/// Announcement subroutine.
+async fn announce(state: State, store: kv::Store, mut sender: mpsc::Sender<Input>) {
+    match announcement::run(&state, &store).await {
+        Ok(updates) => {
+            sender
+                .send(Input::Announce(AnnounceInput::Succeeded(updates)))
+                .await
+                .ok();
+        },
+        Err(err) => {
+            log::error!("announce error: {:?}", err);
+            sender
+                .send(Input::Announce(AnnounceInput::Failed))
+                .await
+                .ok();
+        },
+    }
+}
+
+/// Control response subroutine.
```

Comment:
These docs don't add much since we already have the module and the fn name. as a reader I'd better like to be informed of what these do and when they are useful.